### PR TITLE
[mle] add API to configure the store frame counter ahead

### PR DESCRIPTION
**Pull request by markaj-nordic**
Currently the OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD is hard-coded in the OT core, yet sometimes it is
desired to modify this value which requires
re-building of the OT libraries and forces
re-certification of the end product.

Implement `otThreadSetStoreFrameCounterAhead`
and `otThreadGetStoreFrameCounterAhead` to allow
API clients to configure the store frame counter
ahead parameter at run-time. This extension offloads product makers from the need of re-certification
in case the store frame counter ahead must be tuned.

### CODE REVIEW

**Review by maciejbaczmanski — APPROVED** [review]
LGTM, CLA issue should be resolved soon. Thanks @markaj-nordic!

**Comment by abtink** [comment]
> Currently the OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD is hard-coded in the OT core, yet sometimes it is desired to modify this value which requires re-building of the OT libraries and forces re-certification of the end product.

@markaj-nordic, what scenarios necessitate changing `FRAME_COUNTER_AHEAD` on a product?

One concern about allowing run-time changes to this configuration is the potential for it to set a precedent, leading to requests for similar APIs to modify other `OPENTHREAD_CONFIG` constants. This could result in increased code size and complexity.

**Comment by markaj-nordic** [comment]
> > Currently the OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD is hard-coded in the OT core, yet sometimes it is desired to modify this value which requires re-building of the OT libraries and forces re-certification of the end product.
> 
> @markaj-nordic, what scenarios necessitate changing `FRAME_COUNTER_AHEAD` on a product?
> 
> One concern about allowing run-time changes to this configuration is the potential for it to set a precedent, leading to requests for similar APIs to modify other `OPENTHREAD_CONFIG` constants. This could result in increased code size and complexity.

@abtink I get your point. This patch aims to address the case, when the product maker needs to increase the frame counter ahead to limit the excessive NVS usage. And it seems like an overkill to force the product to be re-certified just because one parameter was tuned to optimize the device performance in the field.

**Comment by markaj-nordic** [comment]
@abtink, any more feedback here?

**Comment by abtink** [comment]
> @abtink, any further feedback on this?

I'm open to adding this API if it's genuinely needed, but I'd like to explore whether alternative solutions might be viable.

Specifically, I'm wondering:

* **Use Case:** How and why would a product maker need to adjust this value? 
* **Default Value:** Could we potentially set the `OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD` config to a sufficiently large value that would work well in various scenarios, eliminating the need for tuning per product?

**Comment by markaj-nordic** [comment]
> > @abtink, any further feedback on this?
> 
> I'm open to adding this API if it's genuinely needed, but I'd like to explore whether alternative solutions might be viable.
> 
> Specifically, I'm wondering:
> 
>     * **Use Case:** How and why would a product maker need to adjust this value?
> 
>     * **Default Value:** Could we potentially set the `OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD` config to a sufficiently large value that would work well in various scenarios, eliminating the need for tuning per product?

* **Use Case:** Depending on the persistent storage implementation in use, the default value of this parameter may lead to some edge cases with garbage collector that removes non-recent key-value pairs records. For our particular use case, with 8k of Zephyr settings partition, after 50 writes to NVM, the garbage collection is triggered because there is not enough margin left to the flash page size of 4096 Bytes. Another obvious reason for decreasing the `OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD`  is limitation of operations on flash, i.e to reduce wear-out. The`OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD` is an internal OT config, therefore it cannot be changed for prebuilt libraries leaving the customers with the need of recertification if they find such a problem.
* **Default Value:** When it comes to a default value, we have a customer that is going to use a value of 100K (multiply the existing one by 100). But I cannot tell for sure if this will satisfy the others as well - it may e.g. depend on the persistent storage in use (more specifically for Zephyr - the size of the settings partition and garbage collector implementation)

All in all, seems that the desired value of `OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD` is quite application specific.

File: include/openthread/thread.h [hunk 1]
```diff
@@ -1158,6 +1158,25 @@ otError otThreadDetachGracefully(otInstance *aInstance, otDetachGracefullyCallba
  */
 void otConvertDurationInSecondsToString(uint32_t aDuration, char *aBuffer, uint16_t aSize);
 
+/**
+ * Sets the store frame counter ahead.
+ *
```
Comment (abtink) [inline include/openthread/thread.h:1163]: May help to document what the "frame counter ahead" parameter is to help others reading the API to underatand why/how to use it and that is should be generally configued using the `OPENTHREAD_CONFIG_STORE_FRAME_COUNTER_AHEAD`

Maybe something like:

The OpenThread stack stores the MLE and MAC security frame counter values in non-volatile storage, ensuring they persist across device resets. These saved values are set to be ahead of their current values by the "frame counter ahead" value.

**Comment by markaj-nordic** [comment]
@abtink thanks for the feedback and comments.

> If there's a change in the platform implementation...

> The same logic would apply to changes in the non-volatile settings platform implementation...

I'd say it's the other way around - it's the platform that is being to some extend affected by the high-level configuration of the OT, in this case `OPENTHEAD_CONFIG_STORE_FRAME_COUNTER_AHEAD` parameter.

Anyway, I think that the default value of the `OPENTHEAD_CONFIG_STORE_FRAME_COUNTER_AHEAD` can be increased up to 100k. I did some math and it seems that this value fits well. Even if we take into account the sequence/frame counter behavior and assume extreme use cases like:

* the device is rebooted extremely frequently, i.e. each 10 minutes
* there is a TX frame each and every 10 ms

we will not reach the max of 32-bit frame counter.

I guess we can tune this in the platform code, as you said, but maybe we should consider introducing such default value to the global OT configuration as well?

But also, keeping in mind above, isn't the frequency of the device reboot and TX frame pattern something product specific? :) Maybe that could justify the API change?

File: include/openthread/thread.h (hunk 1 above)
Comment (markaj-nordic) [reply include/openthread/thread.h:1163]: Good point, thanks.

**Comment by markaj-nordic** [comment]
@abtink, any further remarks?

**Comment by abtink** [comment]
> I guess we can tune this in the platform code, as you said, but maybe we should consider introducing such a default value to the global OT configuration as well?

We can consider changing the default value, though the platform would be the best place to decide this.

> But also, keeping in mind the above, isn't the frequency of the device reboot and TX frame pattern something product-specific? :) Maybe that could justify the API change?

Generally speaking, the platform should aim for reasonable expected worst-case scenarios.

That said, I am not opposed to adding the API for this. Maybe adding the API itself under a config. My main reservation about this is to avoid setting a precedent where folks start adding APIs for other configurations.

**Review by maciejbaczmanski — COMMENTED** [review]
@markaj-nordic I think adding the `OPENTHREAD_CONFIG_DYNAMIC_STORE_FRAME_AHEAD_COUNTER_ENABLE` to `options.cmake` might make it easier for vendors to configure this functionality

File: src/core/thread/mle.hpp [hunk 2]
```diff
@@ -737,6 +737,25 @@ class Mle : public InstanceLocator, private NonCopyable
...
+    }
+
+    /**
+     * Gets the current store frame counter ahead.
+     *
+     * @returns The current store frame counter ahead.
+     *
+     */
+    uint32_t GetStoreFrameCounterAhead() { return mStoreFrameCounterAhead; }
+
```
[the remark this answers is not in the document]
Comment (markaj-nordic) [reply src/core/thread/mle.hpp]: Done, thanks.

**Comment by markaj-nordic** [comment]
> @markaj-nordic I think adding the `OPENTHREAD_CONFIG_DYNAMIC_STORE_FRAME_AHEAD_COUNTER_ENABLE` to `options.cmake` might make it easier for vendors to configure this functionality

@maciejbaczmanski, ok, added this config to options.cmake.

File: etc/cmake/options.cmake [hunk 3]
```diff
@@ -255,6 +255,7 @@ ot_option(OT_TX_QUEUE_STATS OPENTHREAD_CONFIG_TX_QUEUE_STATISTICS_ENABLE "tx que
 ot_option(OT_UDP_FORWARD OPENTHREAD_CONFIG_UDP_FORWARD_ENABLE "UDP forward")
 ot_option(OT_UPTIME OPENTHREAD_CONFIG_UPTIME_ENABLE "uptime")
 ot_option(OT_VERHOEFF_CHECKSUM OPENTHREAD_CONFIG_VERHOEFF_CHECKSUM_ENABLE "verhoeff checksum")
+ot_option(OT_DYNAMIC_STORE_FRAME_AHEAD_COUNTER OPENTHREAD_CONFIG_DYNAMIC_STORE_FRAME_AHEAD_COUNTER_ENABLE "dynamic store frame ahead counter")
```
Comment (abtink) [inline etc/cmake/options.cmake:258]: Move this up close to other options starting with `OT_D*`?
We generally aim to keep this list sorted.

File: src/core/thread/mle.hpp [hunk 4]
```diff
@@ -737,6 +737,27 @@ class Mle : public InstanceLocator, private NonCopyable
...
+        mStoreFrameCounterAhead = aStoreFrameCounterAhead;
+    }
+
+    /**
+     * Gets the current store frame counter ahead.
+     *
+     * @returns The current store frame counter ahead.
+     *
+     */
+    uint32_t GetStoreFrameCounterAhead() { return mStoreFrameCounterAhead; }
```
Comment (abtink) [inline src/core/thread/mle.hpp:758]:
```suggestion
    uint32_t GetStoreFrameCounterAhead(void) { return mStoreFrameCounterAhead; }
```

**Review by abtink — APPROVED** [review]
LGTM. Thanks for adding the `OPENTHREAD_CONFIG_DYNAMIC_STORE_FRAME_AHEAD_COUNTER_ENABLE` config.

Couple of smaller suggestions below:

File: etc/cmake/options.cmake (hunk 3 above)
Comment (markaj-nordic) [reply etc/cmake/options.cmake]: Sure, done.

File: src/core/thread/mle.hpp (hunk 4 above)
Comment (markaj-nordic) [reply src/core/thread/mle.hpp]: Thanks for the catch, fixed.

**Comment by markaj-nordic** [comment]
@abtink I have addressed your last comments. I'd appreciate if you could start the workflows.

**Review by Vyrastas — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**